<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откроем-файл-с-данными-и-изучим-его." data-toc-modified-id="Откроем-файл-с-данными-и-изучим-его.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откроем файл с данными и изучим его.</a></span></li><li><span><a href="#Разделим-исходные-данные-на-обучающую,-валидационную-и-тестовую-выборки." data-toc-modified-id="Разделим-исходные-данные-на-обучающую,-валидационную-и-тестовую-выборки.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разделим исходные данные на обучающую, валидационную и тестовую выборки.</a></span></li><li><span><a href="#Исследуем-качество-разных-моделей,-меняя-гиперпараметры." data-toc-modified-id="Исследуем-качество-разных-моделей,-меняя-гиперпараметры.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуем качество разных моделей, меняя гиперпараметры.</a></span><ul class="toc-item"><li><span><a href="#Модель-дерева-решений.-DecisionTreeClassifier" data-toc-modified-id="Модель-дерева-решений.-DecisionTreeClassifier-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Модель дерева решений. DecisionTreeClassifier</a></span></li><li><span><a href="#Модель-случайный-лес.-RandomForestClassifier" data-toc-modified-id="Модель-случайный-лес.-RandomForestClassifier-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Модель случайный лес. RandomForestClassifier</a></span></li><li><span><a href="#Логистическая-регрессия.LogisticRegression" data-toc-modified-id="Логистическая-регрессия.LogisticRegression-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Логистическая регрессия.LogisticRegression</a></span></li><li><span><a href="#Вывод." data-toc-modified-id="Вывод.-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод.</a></span></li></ul></li><li><span><a href="#Проверим-качество-модели-на-тестовой-выборке" data-toc-modified-id="Проверим-качество-модели-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверим качество модели на тестовой выборке</a></span></li><li><span><a href="#Проверим-на-вменяемость" data-toc-modified-id="Проверим-на-вменяемость-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверим на вменяемость</a></span></li><li><span><a href="#Общий-вывод." data-toc-modified-id="Общий-вывод.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общий вывод.</a></span></li></ul></div>

# Описание проекта , цель

Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Они хотят построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».

**Цель:** Постройть модель с максимально большим значением accuracy. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте accuracy на тестовой выборке самостоятельно.

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
from sklearn.model_selection import train_test_split

In [4]:
from sklearn.tree import DecisionTreeClassifier

In [5]:
from sklearn.metrics import accuracy_score

In [6]:
from sklearn.ensemble import RandomForestClassifier

In [7]:
from sklearn.linear_model import LogisticRegression

<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что собираешь все импорты в верхней части ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
  
</div>

## Откроем файл с данными и изучим его.

In [8]:
df = pd.read_csv('/datasets/users_behavior.csv') #/datasets/users_behavior.csv

In [9]:
df.shape

(3214, 5)

In [10]:
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Согласно условию, и первому взгляду на датафрейм , предобработку данных делать не надо.Пропушеных значений нет.Это важно для машинного обучения.

## Разделим исходные данные на обучающую, валидационную и тестовую выборки.

Согласно теории (как нас учили) данные необходимо разделить на доли 3:1:1,где 3 доля обучающих данных, а по одному доля валидационых и тестовых данных. Для начала разделим весь датафрейм на основной(его в свою очередь будем делить на обучающий и валидационный) и тестовый.

In [12]:
main_df,test_df = train_test_split(df,test_size=0.20,random_state=12345)

In [13]:
main_df.shape

(2571, 5)

In [14]:
test_df.shape

(643, 5)

Из main_df мы выдели features и targets  разделим на обущающие и валидационные.
В нашей модели признаками(features) будут все столбцы кроме столбца `is_ultra`

In [15]:
features = main_df.drop('is_ultra',axis=1)

In [16]:
features.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2571 entries, 348 to 482
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     2571 non-null   float64
 1   minutes   2571 non-null   float64
 2   messages  2571 non-null   float64
 3   mb_used   2571 non-null   float64
dtypes: float64(4)
memory usage: 100.4 KB


А целею(targets)только столбец `is_ultra`.

In [17]:
targets=main_df['is_ultra']

Поделим наши признаки и цели на `обучающие(train)` и `валидацинные данные(valid)`

In [18]:
features_train, features_valid, target_train, target_valid = train_test_split(features,targets,test_size=0.25, random_state=12345)

In [19]:
features_train.shape

(1928, 4)

In [20]:
features_valid.shape

(643, 4)

Данные готовы для обучения моделей , тестовую выборку поделим на признаки и цель перед проверкой модели.

<div class="alert alert-block alert-success">
    
<b>Успех:</b> С разбиением все ок, молодец, что не забываешь проверять полученные выборки,  это иной раз может предостеречь тебя от глупых ошибок.
  </div>  



## Исследуем качество разных моделей, меняя гиперпараметры.

### Модель дерева решений. DecisionTreeClassifier

Применим модель DecisionTreeClassifier.Будем изменять параметр `max_depth` с помьющью цикла и определим его. Проверять `модель(model)` на данных валидационной выборки. Сделаем предсказание `predictions` и проверём точность модели на `target_valid` с помощью `accuracy_score()`

In [21]:
best_model = None
best_result = 0
for depth in range(1, 7):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model.fit(features_train,target_train) 
    predictions = model.predict(features_valid) 
    result = accuracy_score(target_valid,predictions )
    if result > best_result:
        best_model = model
        best_result = result
        best_depth = depth
        
print("Accuracy лучшей модели:", best_result)
print("Лучшее значение параметра depth;",best_depth)

Accuracy лучшей модели: 0.7651632970451011
Лучшее значение параметра depth; 3


Наша точность в принципе  соответсвует  цели, но будем пробывать другие модели дабы получить более лучшей результат. 

### Модель случайный лес. RandomForestClassifier

Применим модель "случайный лес" и будем менять такие параметры как `n_estimators` и `max_depth`

In [22]:
best_model = None
best_result = 0
for est in range(1, 100,10):
    for depth in range(10,20):
        model = RandomForestClassifier(random_state=12345, n_estimators=est,max_depth=depth) 
        model.fit(features_train,target_train)
        predictions = model.predict(features_valid)
        result = accuracy_score(target_valid,predictions)
        if result > best_result:
            best_model = model
            best_result = result
            best_est = est
            best_depth = depth
print("Accuracy лучшей модели:", best_result)
print("Лучшее значение параметра n_estimators=est:",best_est)
print("Лучшее значение параметра depth;",best_depth)

Accuracy лучшей модели: 0.8009331259720062
Лучшее значение параметра n_estimators=est: 21
Лучшее значение параметра depth; 15


Точность получись лучше чем в первой модели, но процесс вычисления был долгим. Если бы данных было бы больше я думаю было бы ещё дольше. Такжк на скорость повлияло то, что мы использовали ёще циклы для перебора параметров.

### Логистическая регрессия.LogisticRegression

In [23]:
best_result = 0
for ite in range(100,2000,100):
    model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=ite)
    model.fit(features_train,target_train)
    predictions = model.predict(features_valid)
    result = accuracy_score(target_valid,predictions)
    if result > best_result:
        best_result = result
        best_iter= ite
print("Accuracy лучшей модели:",best_result)
print("Лучшее значение параметра max_iter;",best_iter)

Accuracy лучшей модели: 0.7262830482115086
Лучшее значение параметра max_iter; 100


Модель логистическая регрессия показала себя наиболее худшей.

### Вывод.

Наилучшее качество модели(0.80093) мы добились используя случайный лес с параметрами n_estimators=21, max_depth=15. Провери на тестовой выборке.

## Проверим качество модели на тестовой выборке

Для начала наши тестовые данные разделим на признаки и цель.

In [24]:
test_features = test_df.drop('is_ultra',axis=1)

In [25]:
test_features.shape

(643, 4)

In [26]:
test_target = test_df['is_ultra']

Переобучим нашу модель снова и проверим на точность на тестовой выборке.

In [27]:
model = RandomForestClassifier(random_state=12345, n_estimators=21,max_depth=15) 
model.fit(features_train,target_train)
predictions = model.predict(test_features )
result = accuracy_score(test_target,predictions)
print("Точность модели на тестовой выборке:",result)

Точность модели на тестовой выборке: 0.7822706065318819


## Проверим на вменяемость 

Чтобы проверить модель на вменяемость сгенирируем последовательность с заролненую единицами и посчитаем
точность . Посути мы посчитаем долю единиц(средняя арифметическая) в нашем predictions.

Посчитаем длину target

In [52]:
len(test_target)

643

Создаём Series с единицами.

In [53]:
t_1=pd.Series(np.ones(643))

In [54]:
result = accuracy_score(t_1,predictions)
print("Точность модели на тестовой выборке:",result)

Точность модели на тестовой выборке: 0.22706065318818042


<div class="alert alert-block alert-success">

<b>Успех:</b> Лучше было бы попробовать нули

Сравним со средим predictions

In [55]:
predictions.mean()

0.22706065318818042

Совпадает со средним.

Также можно сгенировать биноманальную последоватльность случайную с заданой вероятность 0.5 и сравнить с 
predictions. Должно получиться число приближенное к 0.5

In [59]:
t_2 = pd.Series(np.random.binomial(1,0.5 ,size=643))

In [60]:
result = accuracy_score(t_2,predictions)
print("Точность модели на тестовой выборке:",result)

Точность модели на тестовой выборке: 0.49144634525660963


Это говорить что наша модель адекватная.

<div class="alert alert-block alert-success">

<b>Успех:</b> Бонусное задание выполнено корректно, советую также обратить внимание на DummyClasssifier

## Общий вывод.

**Более качественная модель "случайный лес" RandomForestClassifier (random_state=12345, n_estimators=21,max_depth=15), выдаёт точность 0.7822 на тестовой выборке. Проверерку на вменяемость 
прошла **

<div class="alert alert-block alert-success">

<b>Успех:</b> Всегда приятно наблюдать итоговый вывод в конце работы